contracts/vault/CurveSwapper.sol

```diff
@@ -0,0 +1,175 @@
+        require(
+            _tokens.length == _underlyingIs.length,
+            "invalid _underlyingsIs length"
+        );
+        require(
+            address(_underlying) != address(0),
+            "_underlying cannot be 0x0"
+        );
+
+        underlying = _underlying;
+
+        for (uint256 i; i < _tokens.length; ++i) {
+            addPool(_tokens[i], _pools[i], _tokenIs[i], _underlyingIs[i]);
+        }
+    }
+
+    function addPool(
+        address _token,
+        ICurve _pool,
+        int128 _tokenI,
+        int128 _underlyingI
+    ) public {
+        require(
+            _pool.coins(uint256(uint128(_underlyingI))) == underlying,
+            "_underlyingI does not match underlying token"
+        );
+
+        // TODO it seems this doesn't actually work for UST-3CRV
+        // require(
+        //     _pool.coins(uint256(uint128(_tokenI))) == _token,
+        //     "_tokenI does not match input token"
+        // );
+
+        uint256 tokenDecimals = IERC20Metadata(_token).decimals();
+        uint256 underlyingDecimals = IERC20Metadata(underlying).decimals();
+
+        // TODO check if _token and _underlyingIndex match the pool settings
+        swappers[_token] = Swapper(
+            _pool,
+            uint8(tokenDecimals),
+            uint8(underlyingDecimals),
+            _tokenI,
+            _underlyingI
+        );
+
+        _approveIfNecessary(underlying, address(_pool));
+        _approveIfNecessary(_token, address(_pool));
+    }
+
+    /// Swaps a given amount of
+    /// Only works if the pool has previously been inserted into the contract
+    ///
+    /// @param _token The token we want to swap into
+    /// @param _amount The amount of underlying we want to swap
+    /// TODO missing slippage checks
+    function _swapIntoUnderlying(address _token, uint256 _amount) internal {
+        if (_token == underlying) {
+            // same token, nothing to do
+            return;
+        }
+
+        Swapper storage swapper = swappers[_token];
+
+        uint256 minAmount = _calc_dy(
+            _amount,
+            swapper.tokenDecimals,
+            swapper.underlyingDecimals
+        );
+
+        swapper.pool.exchange_underlying(
+            swapper.tokenI,
+            swapper.underlyingI,
+            _amount,
+            minAmount
+        );
+    }
+
+    /// Swaps a given amount of Underlying into a given token
+    /// Only works if the pool has previously been inserted into the contract
+    ///
+    /// @param _token The token we want to swap into
+    /// @param _amount The amount of underlying we want to swap
+    /// TODO missing slippage checks
+    function _swapFromUnderlying(address _token, uint256 _amount) internal {
+        if (_token == underlying) {
+            // same token, nothing to do
+            return;
+        }
+
+        Swapper storage swapper = swappers[_token];
+
+        uint256 minAmount = _calc_dy(
```

Comment:
why are you using snakecase for this one? I guess also that this is `min_dy` on the docs right?

Reply:
I used snakeCase purely because curve uses it for the argument name. but yes, I can change and also include `min`